### PR TITLE
REDBOX-396: user ratings

### PR DESCRIPTION
**Pull request by brunns**
## Context

Collect user ratings & make available for export - see https://technologyprogramme.atlassian.net/browse/REDBOX-396

## Changes proposed in this pull request

Screenshot 2024-07-11 at 11 42 33
Screenshot 2024-07-11 at 11 42 54
Screenshot 2024-07-11 at 11 43 10
Screenshot 2024-07-11 at 11 43 19
[chathistory.csv](https://github.com/user-attachments/files/16175612/chathistory.csv)

## Things to check

- [ ] I have added any new ENV vars in all deployed environments
- [x] I have tested any code added or changed
- [x] I have run integration tests

### CODE REVIEW

File: django_app/frontend/js/chats/feedback.js [hunk 1]
```diff
@@ -1,43 +1,146 @@
...
+                <path class="rb-response-feedback__thumb-fill" d="M 22.088 10.651 z M 6 22 H 3 V 11 h 3 z m 14.571 -9.251 l -0.582 0.363 l 0.525 0.443 a 1.27 1.27 0 0 1 0.508 1.175 a 1.359 1.359 0 0 1 -0.892 1.013 l -0.747 0.305 l 0.604 0.533 a 1.208 1.208 0 0 1 0.395 1.227 a 1.167 1.167 0 0 1 -0.908 0.851 l -0.775 0.167 l 0.485 0.628 a 0.858 0.858 0 0 1 0.153 0.939 a 1.25 1.25 0 0 1 -1.148 0.607 h -6.646 l -2.472 -0.966 L 7 20.007 V 13 h 1.473 l 2.448 -3.395 a 9.933 9.933 0 0 0 2.683 -2.867 a 8.134 8.134 0 0 0 1.328 -3.772 a 0.654 0.654 0 0 1 0.562 -0.089 c 0.166 0.078 0.708 0.52 0.708 2.955 a 4.09 4.09 0 0 1 -1.101 2.614 a 1.051 1.051 0 0 0 -0.237 1.06 c 0.25 0.494 0.87 0.494 1.896 0.494 h 3.391 c 0.524 0 0.847 0.48 0.976 0.928 a 1.616 1.616 0 0 1 -0.556 1.821 z"/>    
+            </svg>
+            <span class="feedback__negative">Sorry this didn't meed your expectations</span>
```
Comment (brunns) [inline django_app/frontend/js/chats/feedback.js:50]:
```suggestion
            <span class="feedback__negative">Sorry this didn't meet your expectations</span>
```

File: django_app/redbox_app/redbox_core/models.py [hunk 2]
```diff
@@ -255,3 +256,23 @@ class ChatMessage(UUIDPrimaryKeyBase, TimeStampedModel):
 
     def __str__(self) -> str:  # pragma: no cover
         return f"{self.text} - {self.role}"
+
+
+class ChatMessageRating(TimeStampedModel):
+    chat_message = models.OneToOneField(ChatMessage, on_delete=models.CASCADE, primary_key=True)
+    rating = models.PositiveIntegerField(validators=[validators.MinValueValidator(1), validators.MaxValueValidator(5)])
```
Comment (rachaelcodes) [inline django_app/redbox_app/redbox_core/models.py:263]: Do we need to account for a future where the range of options changes? Even if we update these validators, will we be able to make sure that ratings gathered when they are out of 5 are comparable to ones gathered when it's out of 10 (for example)?

Comment (brunns) [reply django_app/redbox_app/redbox_core/models.py:263]: We would need to fiddle with the migrations if the range changes - so, doubling all existing ratings, given our examples. Easy enough to do, I think.

File: django_app/frontend/src/styles.scss [hunk 3]
```diff
@@ -170,82 +170,167 @@ body {
...
+  color: black;
+}
+.feedback__chip-label img {
+  display: none;
+  width: 0.75rem;
+}
+.feedback__chip:checked + .feedback__chip-label img {
+  display: block;
+}
+.feedback__text-input {
```
Comment (rachaelcodes) [inline django_app/frontend/src/styles.scss:317]: Should this be `resize: none;`? The resizing handle makes it look notably different from the other textareas. 
Screenshot 2024-07-11 at 12 55 07

Comment (brunns) [reply django_app/frontend/src/styles.scss:317]: @KevinEtchells ?

Comment (KevinEtchells) [reply django_app/frontend/src/styles.scss:317]: As it auto-resizes, that sounds like a good idea

File: django_app/redbox_app/redbox_core/admin.py [hunk 4]
```diff
@@ -86,21 +87,34 @@ class ChatMessageInline(admin.StackedInline):
...
-        writer.writerow(history_field_names + message_field_names)
+        writer.writerow(
+            ["history_" + n for n in history_field_names]
+            + ["message_" + n for n in message_field_names]
+            + ["rating_" + n for n in rating_field_names]
+            + ["rating_chips"]
+        )
+        chat_history: models.ChatHistory
+        chat_message: models.ChatMessage
+        chat_massage_rating: models.ChatMessageRating
```
Comment (rachaelcodes) [inline django_app/redbox_app/redbox_core/admin.py:107]: think this is a typo?

Comment (brunns) [reply django_app/redbox_app/redbox_core/admin.py:107]: I'm going to pretend it's a deliberate Monty Python reference - https://www.youtube.com/watch?v=P75lkpL7aGk - but I will fix.

File: django_app/tests/test_admin.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+
+@pytest.mark.django_db()
+def test_chat_history_export(superuser: User, chat_message_with_rating: ChatMessage, client: Client):
+    # Given
+    client.force_login(superuser)
+
+    # When
+
+    # See https://docs.djangoproject.com/en/dev/ref/contrib/admin/#reversing-admin-urls
```
Comment (rachaelcodes) [inline django_app/tests/test_admin.py:22]: v helpful comment 👍

**Review by rachaelcodes — APPROVED** [review]
LGTM - this is an epic PR and looks great!

**Pull request merged**